File: app/api/v1/portfolios.py
import uuid

from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from app.dependencies import get_current_user_id, get_db
from app.domain.portfolio import service
from app.domain.portfolio.schemas import PortfolioCreate, PortfolioResponse

router = APIRouter(prefix="/portfolios", tags=["portfolios"])


@router.get("", response_model=list[PortfolioResponse])
async def list_portfolios(
    user_id: uuid.UUID = Depends(get_current_user_id),
    session: AsyncSession = Depends(get_db),
):
    return await service.list_portfolios(session, user_id)


@router.post("", response_model=PortfolioResponse, status_code=201)
async def create_portfolio(
    data: PortfolioCreate,
    user_id: uuid.UUID = Depends(get_current_user_id),
    session: AsyncSession = Depends(get_db),
):
    return await service.create_portfolio(session, user_id, data)


@router.get("/{portfolio_id}", response_model=PortfolioResponse)
async def get_portfolio(
    portfolio_id: uuid.UUID,
    user_id: uuid.UUID = Depends(get_current_user_id),
    session: AsyncSession = Depends(get_db),
):
    return await service.get_portfolio(session, user_id, portfolio_id)


@router.delete("/{portfolio_id}", status_code=204)
async def delete_portfolio(
    portfolio_id: uuid.UUID,
    user_id: uuid.UUID = Depends(get_current_user_id),
    session: AsyncSession = Depends(get_db),
):
    await service.delete_portfolio(session, user_id, portfolio_id)

File: frontend/src/api/portfolios.js
import { api } from './client'

export const listPortfolios = () => api.get('/portfolios')

export const createPortfolio = (name, description) =>
  api.post('/portfolios', { name, description })

export const getPortfolio = (id) => api.get(`/portfolios/${id}`)

export const deletePortfolio = (id) => api.delete(`/portfolios/${id}`)

export const addPosition = (portfolioId, symbol, quantity, averageCost) =>
  api.post(`/portfolios/${portfolioId}/positions`, {
    symbol,
    quantity,
    average_cost: averageCost,
  })

export const deletePosition = (portfolioId, positionId) =>
  api.delete(`/portfolios/${portfolioId}/positions/${positionId}`)
